app/assets/javascripts/i18n-strings.en.js.erb

```diff
@@ -0,0 +1,6 @@
+<% keys = YAML.load_file Rails.root.join('config', 'js_locale_strings.yml') %>
+
+window.LoginGov.I18n.strings.en = {};
+<% keys.each do |key| %>
+  window.LoginGov.I18n.strings.en['<%= ActionController::Base.helpers.j key %>'] = '<%= ActionController::Base.helpers.j I18n.t(key, locale: 'en') %>';
+<% end %>
```

Comment:
Can we make a partial for these? That way if we ever add a 4th locale, we make it less error-prone? We could get rid of the YAML and put that back in the partial if you wanted. Unsure if it clearer to have a YAML or inline, it probably makes linting easier to have a YAML

```erb
<%= render partial: 'i18n_strings.js', locals: { locale: 'en' } %>
```


In the partial
```erb
<%#
locals:
  - locale: current locale to render (ex "en")
%>
<% keys = YAML.load_file Rails.root.join('config', 'js_locale_strings.yml') %>

window.LoginGov.I18n.strings.<%= locale %>en = {};
<% keys.each do |key| %>
  window.LoginGov.I18n.strings.<%= locale %>['<%= ActionController::Base.helpers.j key %>'] = '<%= ActionController::Base.helpers.j I18n.t(key, locale: locale) %>';
<% end %>
```

Reply:
Yeah, I definitely got similar vibes that the repetition here could be a source of future maintenance headache. Extracting the keys to a configuration file was the most obvious candidate to try to reduce the maintenance overhead. I like your idea of using a partial. Unfortunately, it appears it's [not possible to use `render` from within assets](https://github.com/rails/rails/issues/1370#issuecomment-1303371).

A few alternatives I've been exploring:

- Creating a utility function which builds the string (or loads and evaluates a template file much like `render`) with `locale` as the argument
- As mentioned in the original comment, maybe some way to generate these files dynamically

Reply:
ugh I didn't realize we couldn't use `render` inside assets.... what a mess.

OK new idea: Let's dump the translations in a meta tag as just plain JSON, then we can skip the entirety of the asset pipeline. I believe that `JSON.parse` is faster than eval-ing code (because JSON is a limited subset of JS), so we save a little on client-side parsing time. The backend should be fast-ish as well since it's just in-memory JSONifying of stuff.

base layout:
```erb
<%= tag.meta name: 'js-translations', lang: I18n.locale, content: js_translations.to_json %>
```

translations_helper.rb (or something)
```rb
def js_translations
  keys = [
    'two_factor_authentication.otp_delivery_preference.instruction',
    'two_factor_authentication.otp_delivery_preference.phone_unsupported',
    # ...
  ]
  translations = keys.map do |key|
    [ key, I18n.t(key) ]
  end

  { I18n.locale => translations.to_h }
end
```

i18n.js
```js
window.LoginGov.i18n.backend = JSON.parse(document.querySelector('meta[name=js-translations]').content)
```

Reply:
>The backend should be fast-ish as well since it's just in-memory JSONifying of stuff.

Not sure if it's what you had in mind for performance, but one of the other thoughts I had here, aside from the fact that we'd be generating it on each page request (vs. precompiling a static JS file) is that the JS file can also be cached by the user's browser, whereas if we stuff it into the page meta tags, there'll be no caching.

Would you have any concern about that?

Reply:
Yeah breaking into a separate file so it can be cached is fine! But we're not doing that? It looks like `i18n-strings` is just compiled inside of `application.js` on every file load right now?

Ah but I see the code to make the translations its own pack now.... ok so this all makes sense, I wish the asset pipeline was easier to work with, thanks for walking through all this with me

Reply:
> I wish the asset pipeline was easier to work with

You and me both! 😄  I'm sure there's a better approach here with middlewares, etc. Just feel it's probably a bit more involved an effort than might be worth carving off vs. this simplistic approach.

I did a bit more tinkering to try to reduce the boilerplate some more, toward something like in [your earlier suggestion](https://github.com/18F/identity-idp/pull/4034#discussion_r467986173):

```suggestion
<% keys = YAML.load_file Rails.root.join('config', 'js_locale_strings.yml') %>

window.LoginGov.I18n.strings.en = <%= keys.map { |key| [key, I18n.t(key, locale: 'en')] }.to_h.to_json %>;
```

It's shorter, but doesn't really solve the bigger problem of the two occurrences of the explicit `en` reference.

I'll go ahead and merge this as-is, but if you think above (or equivalent) would be an improvement, I'm happy to open a follow-up pull request.

Reply:
Actually, in the pursuit of optimization, the above has another nice advantage: It doesn't write out the full object name for each assignment.

In other words:

```js
// Before (10.7kb unminified, pre-gzip):
window.LoginGov.I18n.strings.es = {};
  window.LoginGov.I18n.strings.es['doc_auth.buttons.take_picture'] = 'Tomar la foto';
  window.LoginGov.I18n.strings.es['doc_auth.buttons.take_picture_retry'] = 'Retomar foto';
  window.LoginGov.I18n.strings.es['doc_auth.forms.change_file'] = 'Cambiar archivo';
// ...

// After (7.9kb unminified, pre-gzip):
window.LoginGov.I18n.strings.es = {"doc_auth.buttons.take_picture":"Tomar la foto","doc_auth.buttons.take_picture_retry":"Retomar foto","doc_auth.forms.change_file":"Cambiar archivo" /* ... */};
```